### PR TITLE
Generate unique Functions requestIds for reorg'ed requests

### PR DESCRIPTION
**Pull request by KuphJr**

### CODE REVIEW

File: contracts/src/v0.8/functions/dev/v1_0_0/FunctionsBilling.sol [hunk 1]
```diff
@@ -218,26 +211,29 @@ abstract contract FunctionsBilling is Routable, IFunctionsBilling {
...
+
+    s_requestCommitments[commitment.requestId] = keccak256(abi.encode(commitment));
 
     return commitment;
   }
 
   /// @notice Generate a keccak hash request ID
   /// @dev uses the number of requests that the consumer of a subscription has sent as a nonce
   function _computeRequestId(
-    address don,
```
Comment (bolekk) [inline contracts/src/v0.8/functions/dev/v1_0_0/FunctionsBilling.sol:235]: The "don" field was incorrectly named so it's good that it's gone now :)

File: contracts/src/v0.8/functions/dev/v1_0_0/FunctionsBilling.sol [hunk 2]
```diff
@@ -218,26 +211,29 @@ abstract contract FunctionsBilling is Routable, IFunctionsBilling {
       callbackGasLimit: request.callbackGasLimit,
       estimatedTotalCostJuels: estimatedTotalCostJuels,
       timeoutTimestamp: uint32(block.timestamp + config.requestTimeoutSeconds),
-      requestId: requestId,
+      requestId: 0x0000000000000000000000000000000000000000000000000000000000000000,
```
Comment (bolekk) [inline contracts/src/v0.8/functions/dev/v1_0_0/FunctionsBilling.sol:214]: You don't have to set it, do you? Just remove and leave a comment that ID is left unset on purpose and set later.

File: contracts/src/v0.8/functions/dev/v1_0_0/FunctionsBilling.sol [hunk 3]
```diff
@@ -218,26 +211,29 @@ abstract contract FunctionsBilling is Routable, IFunctionsBilling {
...
   /// @dev uses the number of requests that the consumer of a subscription has sent as a nonce
   function _computeRequestId(
-    address don,
-    address client,
-    uint64 subscriptionId,
+    FunctionsResponse.Commitment memory commitmentWithoutRequestId,
     uint64 nonce
   ) private pure returns (bytes32) {
-    return keccak256(abi.encode(don, client, subscriptionId, nonce));
+    return keccak256(abi.encode(commitmentWithoutRequestId, nonce));
```
Comment (bolekk) [inline contracts/src/v0.8/functions/dev/v1_0_0/FunctionsBilling.sol:236]: Is there any meaningful gas overhead of hashing the whole struct?

File: contracts/src/v0.8/functions/dev/v1_0_0/FunctionsBilling.sol [hunk 4]
```diff
@@ -203,13 +203,6 @@ abstract contract FunctionsBilling is Routable, IFunctionsBilling {
       revert InsufficientBalance();
     }
 
-    bytes32 requestId = _computeRequestId(
```
Comment (bolekk) [inline contracts/src/v0.8/functions/dev/v1_0_0/FunctionsBilling.sol:206]: Let's do contract changes in a separate PR for easier tracking/review.

File: core/services/functions/models.go [hunk 5]
```diff
@@ -111,7 +111,7 @@ func CheckStateTransition(prev RequestState, next RequestState) error {
 			IN_PROGRESS:  errors.New("cannot go back from FINALIZED to IN_PROGRESS"),
 			RESULT_READY: errors.New("cannot go back from FINALIZED to RESULT_READY, result was already finalized by DON before this request was picked up"),
 			TIMED_OUT:    nil, // timed out while in transmission - no reason to attempt sending it any more
-			FINALIZED:    sameStateError,
+			FINALIZED:    nil, // the logPoller may have pick up the same response log twice, but it's already finalized
```
Comment (bolekk) [inline core/services/functions/models.go:114]: Maybe we can avoid that? See another comment below.

File: core/services/relay/evm/functions/logpoller_wrapper.go [hunk 6]
```diff
@@ -116,16 +110,16 @@ func (l *logPollerWrapper) LatestEvents() ([]evmRelayTypes.OracleRequest, []evmR
...
 		coordinators = append(coordinators, l.proposedCoordinator)
 	}
-	nextBlock := l.nextBlock
 	latest, err := l.logPoller.LatestBlock()
 	if err != nil {
 		l.mu.Unlock()
 		return nil, nil, err
 	}
-	latest -= l.blockOffset
-	if latest >= nextBlock {
```
Comment (bolekk) [inline core/services/relay/evm/functions/logpoller_wrapper.go:126]: What if LatestBlock() is stuck on the same number for a long time due to RPC client problems? Do we want to constantly re-query the same range?

File: core/services/relay/evm/functions/logpoller_wrapper.go [hunk 7]
```diff
@@ -135,20 +129,16 @@ func (l *logPollerWrapper) LatestEvents() ([]evmRelayTypes.OracleRequest, []evmR
...
 		return resultsReq, resultsResp, errors.New("no non-zero coordinators to check")
 	}
-	if latest < nextBlock {
-		l.lggr.Debugw("LatestEvents: no new blocks to check", "latest", latest, "nextBlock", nextBlock)
-		return resultsReq, resultsResp, nil
-	}
 
 	for _, coordinator := range coordinators {
-		requestLogs, err := l.logPoller.Logs(nextBlock, latest, functions_coordinator.FunctionsCoordinatorOracleRequest{}.Topic(), coordinator)
+		requestLogs, err := l.logPoller.Logs(startBlock, latest, functions_coordinator.FunctionsCoordinatorOracleRequest{}.Topic(), coordinator)
```
Comment (bolekk) [inline core/services/relay/evm/functions/logpoller_wrapper.go:134]: I don't like that we're relying on pluginORM.CreateRequest() to filter out duplicates:
- we will be returning potentially lots of data from LatestEvents() for no good reason.
- we lose the ability to differentiate from an actual buggy duplicate because you disabled a log in handleOracleRequestV1().

Can LogPollerWrapper still keep local state of returned IDs and not return the same one twice?

**Review by bolekk — CHANGES_REQUESTED** [review]

File: contracts/src/v0.8/functions/dev/v1_0_0/FunctionsBilling.sol (hunk 4 above)
Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/v1_0_0/FunctionsBilling.sol:206]: Unless you feel strongly about this, I'd prefer to keep this entire change in a single PR since the changes are dependent on each other.

File: contracts/src/v0.8/functions/dev/v1_0_0/FunctionsBilling.sol (hunk 2 above)
Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/v1_0_0/FunctionsBilling.sol:214]: Turns out you do need to specify else Solidity gives a compilation error.

File: contracts/src/v0.8/functions/tests/v1_0_0/FunctionsRouter.t.sol [hunk 8]
```diff
@@ -1137,7 +1114,7 @@ contract FunctionsRouter_Fulfill is FunctionsClientRequestSetup {
       ) // TODO: build this programatically
     });
 
-    _reportAndStore(requestNumberKeys, results, errors, NOP_TRANSMITTER_ADDRESS_1, true, 1);
+    _reportAndStore(requestNumberKeys, results, errors, NOP_TRANSMITTER_ADDRESS_1, false, 1);
```
Comment (KuphJr) [inline contracts/src/v0.8/functions/tests/v1_0_0/FunctionsRouter.t.sol:1117]: @justinkaseman I need help to understand this `expectedToSucceed` argument in `_reportAndStore`.  I am not entirely sure why I had to adjust this value to get tests to pass after I made this minor change to how the `requestId` is generated.

I see that all it seems to do is [record the profit amounts](https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.8/functions/tests/v1_0_0/Setup.t.sol#L443).  However it appears that this [consumes all the logs](https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.8/functions/tests/v1_0_0/Setup.t.sol#L446) and it causes the [vm.expectEmit](https://github.com/smartcontractkit/chainlink/pull/10891/files#diff-5152fdbb2dac463e32e8e79ba3254a378b51e21693519523ff0c12a6d73b90b8L1126) to fail.

I am thinking this might be a simple semantic fix.  For example, instead of naming the argument `expectedToSucceed`, we could call it `recordCostAndProfit` or something.

Does this make sense or am I missing something?

Comment (KuphJr) [reply contracts/src/v0.8/functions/tests/v1_0_0/FunctionsRouter.t.sol]: I figured this out.  I mistakenly left a `vm` cheat code which broke something. All tests now pass

File: contracts/src/v0.8/functions/dev/v1_0_0/FunctionsBilling.sol (hunk 3 above)
Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/v1_0_0/FunctionsBilling.sol:233]: Actually... There shouldn't need to be any changes to the gas overhead since this computation only occurs during the request transaction.  We never use `_computeRequestId` anywhere else (I checked).

**Comment by KuphJr** [comment]
This change added 7348 gas to the request transaction. Not too bad.

File: contracts/src/v0.8/functions/dev/v1_0_0/FunctionsBilling.sol [hunk 9]
```diff
@@ -218,26 +211,26 @@ abstract contract FunctionsBilling is Routable, IFunctionsBilling {
...
+      requestId: bytes32(0), // requestId is computed after hashing the commitment data
       donFee: donFee,
       gasOverheadBeforeCallback: config.gasOverheadBeforeCallback,
       gasOverheadAfterCallback: config.gasOverheadAfterCallback
     });
 
-    s_requestCommitments[requestId] = keccak256(abi.encode(commitment));
+    commitment.requestId = _computeRequestId(commitment, request.initiatedRequests + 1);
+
+    s_requestCommitments[commitment.requestId] = keccak256(abi.encode(commitment));
```
Comment (RensR) [inline contracts/src/v0.8/functions/dev/v1_0_0/FunctionsBilling.sol:222]: Why do you need a different value here than the request ID? You now do a very expensive abi encode twice.

Comment (RensR) [reply contracts/src/v0.8/functions/dev/v1_0_0/FunctionsBilling.sol:222]: in CCIP we use very similar logic, but we don't abi.encode and hash again. We use the request ID as identifier throughout the code. Hashing it again with the commitment doesn't give any additional security either

Comment (KuphJr) [reply contracts/src/v0.8/functions/dev/v1_0_0/FunctionsBilling.sol:222]: Thanks!  We wanted to hash the whole commitment to ensure if any commitment data changed that the requestId would be unique, but we discussed an alternate approach which should provide the same guarantees without wasting gas.

File: contracts/gas-snapshots/functions.gas-snapshot [hunk 10]
```diff
@@ -52,23 +52,23 @@ FunctionsRouter_ProposeContractsUpdate:test_ProposeContractsUpdate_RevertIfNotNe
...
+FunctionsRouter_SendRequest:test_SendRequest_RevertIfDuplicateRequestId() (gas: 192172)
 FunctionsRouter_SendRequest:test_SendRequest_RevertIfEmptyData() (gas: 29405)
 FunctionsRouter_SendRequest:test_SendRequest_RevertIfIncorrectDonId() (gas: 57926)
 FunctionsRouter_SendRequest:test_SendRequest_RevertIfInsufficientSubscriptionBalance() (gas: 185987)
 FunctionsRouter_SendRequest:test_SendRequest_RevertIfInvalidCallbackGasLimit() (gas: 50902)
 FunctionsRouter_SendRequest:test_SendRequest_RevertIfInvalidDonId() (gas: 25061)
 FunctionsRouter_SendRequest:test_SendRequest_RevertIfNoSubscription() (gas: 29111)
 FunctionsRouter_SendRequest:test_SendRequest_RevertIfPaused() (gas: 34247)
-FunctionsRouter_SendRequest:test_SendRequest_Success() (gas: 197669)
+FunctionsRouter_SendRequest:test_SendRequest_Success() (gas: 284316)
```
Comment (bolekk) [inline contracts/gas-snapshots/functions.gas-snapshot:63]: This is big change! Is it just caused by the extra hash/encoding...?

Comment (KuphJr) [reply contracts/gas-snapshots/functions.gas-snapshot:63]: No, it is caused by the slight change to test logic itself. These gas figures are for the cost to run the Foundry test, not just the cost of our contract function calls.

**Review by justinkaseman — APPROVED** [review]
LGTM. From 7k gas to 50 :+1:

File: contracts/src/v0.8/functions/dev/v1_0_0/FunctionsBilling.sol [hunk 11]
```diff
@@ -203,11 +203,16 @@ abstract contract FunctionsBilling is Routable, IFunctionsBilling {
...
-      request.initiatedRequests + 1
+    uint32 timeoutTimestamp = uint32(block.timestamp + config.requestTimeoutSeconds);
+    bytes32 requestId = keccak256(
+      abi.encode(
+        address(this),
+        request.requestingContract,
+        request.subscriptionId,
+        request.initiatedRequests + 1,
+        estimatedTotalCostJuels,
+        timeoutTimestamp
```
Comment (bolekk) [inline contracts/src/v0.8/functions/dev/v1_0_0/FunctionsBilling.sol:214]: This shouldn't be a blocker for this PR but I want to point out that there is a scenario where contract config is changed at roughly the same time as a request is built and when these two events get re-orged into a different order then some other fields might not match (adminFee, donFee, overheads).

Can we check how expensive it is to add all the fields here? Are there any downsides?

Comment (bolekk) [reply contracts/src/v0.8/functions/dev/v1_0_0/FunctionsBilling.sol:214]: Also do we know how our Frontend is going to behave with a request ID that changes on re-org? How about our dashboard and alerting?

Comment (bolekk) [reply contracts/src/v0.8/functions/dev/v1_0_0/FunctionsBilling.sol:214]: Based on recent discussion, we'll want to add tx.origin and request.data here as well.

File: contracts/src/v0.8/functions/dev/v1_X/FunctionsBilling.sol [hunk 12]
```diff
@@ -208,11 +208,20 @@ abstract contract FunctionsBilling is Routable, IFunctionsBilling {
...
+        request.requestingContract,
+        request.subscriptionId,
+        request.initiatedRequests + 1,
+        request.data,
+        request.dataVersion,
+        request.callbackGasLimit,
+        estimatedTotalCostJuels,
+        timeoutTimestamp,
+        tx.origin
+      )
```
Comment (bolekk) [inline contracts/src/v0.8/functions/dev/v1_X/FunctionsBilling.sol:224]: Explanation: we want to hash on everything that can change in a re-org:
- all inputs to Router.sendRequest()
- commitment fields that depend on block params (pretty much everything except static config)
- tx.origin to handle two EOAs submitting the exact same request (super unlikely but still)

Comment (bolekk) [reply contracts/src/v0.8/functions/dev/v1_X/FunctionsBilling.sol:225]: @justinkaseman @KuphJr does it make sense?

**Review by bolekk — DISMISSED** [review]

**Review by justinkaseman — APPROVED** [review]

File: contracts/gas-snapshots/functions.gas-snapshot [hunk 13]
```diff
@@ -53,23 +53,23 @@ FunctionsRouter_ProposeContractsUpdate:test_ProposeContractsUpdate_RevertIfNotNe
...
+FunctionsRouter_SendRequest:test_SendRequest_RevertIfDuplicateRequestId() (gas: 192796)
 FunctionsRouter_SendRequest:test_SendRequest_RevertIfEmptyData() (gas: 29405)
 FunctionsRouter_SendRequest:test_SendRequest_RevertIfIncorrectDonId() (gas: 57926)
 FunctionsRouter_SendRequest:test_SendRequest_RevertIfInsufficientSubscriptionBalance() (gas: 186209)
 FunctionsRouter_SendRequest:test_SendRequest_RevertIfInvalidCallbackGasLimit() (gas: 50902)
 FunctionsRouter_SendRequest:test_SendRequest_RevertIfInvalidDonId() (gas: 25061)
 FunctionsRouter_SendRequest:test_SendRequest_RevertIfNoSubscription() (gas: 29111)
 FunctionsRouter_SendRequest:test_SendRequest_RevertIfPaused() (gas: 34247)
-FunctionsRouter_SendRequest:test_SendRequest_Success() (gas: 198072)
+FunctionsRouter_SendRequest:test_SendRequest_Success() (gas: 284999)
```
Comment (KuphJr) [inline contracts/gas-snapshots/functions.gas-snapshot:64]: I know a lot of this additional gas is because of the test changes, but I am curious how much more gas it adds to hash all this data. Anyone have some rough numbers?
